src/librustc_errors/lib.rs

```diff
@@ -337,6 +338,265 @@ struct HandlerInner {
     stashed_diagnostics: FxIndexMap<(Span, StashKey), Diagnostic>,
 }
 
+/// A helper trait to implement methods on HandlerInner.
+/// This trait is necessary because of the following unfortunate
+/// interaction:
+/// 1. The Handler struct calls methods on HandlerInner while holding
+/// open a Lock
+/// 2. Methods on HandlerInner may panic
+/// 3. The global panic handler may attempt to access *the original
+/// Handler instance*, and acquire the lock re-entrantly.
+///
+/// We resolve this by ensuring that LockGuard acquired by Handler
+/// has been dropped before we panic. This ensures that the global panic
+/// handler can acquire the lock.
+///
+/// To avoid needing to move the actual panic! call into the Handler
+/// struct, we use the following helper trait. This trait is implemented
+/// on LockGuard<HandlerInner>, which allows it to take
+/// a LockGuard by-value. When we want to panic, we first call drop(self),
+/// which releases the lock.
+///
+/// Note that we implement all of these methods as defaulted methods
+/// on the trait itself. Since we have `DerefMut<Target = HandlerInner> `
+/// as a supertrait, we can access fields on HandlerInner just as if
+/// we were dirctly implementing them on HandlerInner.
+///
+/// The actual impl for LockGuard<HandlerInner> has no methods -
+/// all of the logic is written in the default methods
+trait HandlerInnerHelper: DerefMut<Target = HandlerInner> + Sized {
```

Comment:
Instead of using a trait which opens up for overriding, can we use free functions instead for the bits that should not be overridden?